geo/src/algorithm/line_measures/metric_spaces/euclidean/distance.rs

```diff
@@ -0,0 +1,1087 @@
+        Self::distance(*origin, *destination)
+    }
+}
+
+impl<F: CoordFloat> Distance<F, &Point<F>, &Line<F>> for Euclidean {
+    fn distance(origin: &Point<F>, destination: &Line<F>) -> F {
+        geo_types::private_utils::point_line_euclidean_distance(*origin, *destination)
+    }
+}
+
+impl<F: CoordFloat> Distance<F, &Point<F>, &LineString<F>> for Euclidean {
+    fn distance(origin: &Point<F>, destination: &LineString<F>) -> F {
+        geo_types::private_utils::point_line_string_euclidean_distance(*origin, destination)
+    }
+}
+
+impl<F: GeoFloat> Distance<F, &Point<F>, &Polygon<F>> for Euclidean {
+    fn distance(point: &Point<F>, polygon: &Polygon<F>) -> F {
+        // No need to continue if the polygon intersects the point, or is zero-length
+        if polygon.exterior().0.is_empty() || polygon.intersects(point) {
+            return F::zero();
+        }
+        // fold the minimum interior ring distance if any, followed by the exterior
+        // shell distance, returning the minimum of the two distances
+        polygon
+            .interiors()
+            .iter()
+            .map(|ring| Self::distance(point, ring))
+            .fold(Bounded::max_value(), |accum: F, val| accum.min(val))
+            .min(
+                polygon
+                    .exterior()
+                    .lines()
+                    .map(|line| {
+                        ::geo_types::private_utils::line_segment_distance(
+                            point.0, line.start, line.end,
+                        )
+                    })
+                    .fold(Bounded::max_value(), |accum, val| accum.min(val)),
+            )
+    }
+}
+
+// ┌──────────────────────────┐
+// │ Implementations for Line │
+// └──────────────────────────┘
+
+symmetric_distance_impl!(CoordFloat, &Line<F>, Coord<F>);
+symmetric_distance_impl!(CoordFloat, &Line<F>, &Point<F>);
+
+impl<F: GeoFloat> Distance<F, &Line<F>, &Line<F>> for Euclidean {
+    fn distance(line_a: &Line<F>, line_b: &Line<F>) -> F {
+        if line_a.intersects(line_b) {
+            return F::zero();
+        }
+        // minimum of all Point-Line distances
+        Self::distance(&line_a.start_point(), line_b)
+            .min(Self::distance(&line_a.end_point(), line_b))
+            .min(Self::distance(&line_b.start_point(), line_a))
+            .min(Self::distance(&line_b.end_point(), line_a))
+    }
+}
+
+impl<F: GeoFloat> Distance<F, &Line<F>, &LineString<F>> for Euclidean {
+    fn distance(line: &Line<F>, line_string: &LineString<F>) -> F {
+        line_string
+            .lines()
+            .fold(Bounded::max_value(), |acc, segment| {
+                acc.min(Self::distance(line, &segment))
+            })
+    }
+}
+
+impl<F: GeoFloat> Distance<F, &Line<F>, &Polygon<F>> for Euclidean {
+    fn distance(line: &Line<F>, polygon: &Polygon<F>) -> F {
+        if line.intersects(polygon) {
+            return F::zero();
+        }
+
+        // REVIEW: This impl changed slightly.
```

Comment:
Heads up that this impl changed slightly, though the behavior should be the same. I couldn't resist the simplification. 😇 

All the other implementations were simply cut+paste+renamed.

The original impl is here: https://github.com/georust/geo/pull/1232/files#diff-23bef671023e471a3228d4ac73548395b9073026c10fb07f418ac3e81df8cdddL242
```rust
fn euclidean_distance(&self, other: &Polygon<T>) -> T {
  if self.intersects(other) {
    return T::zero();
  }
  // line-line distance between each exterior polygon segment and the line
  let exterior_min = other
    .exterior()
    .lines()
    .fold(<T as Bounded>::max_value(), |acc, point| {
      acc.min(self.euclidean_distance(&point))
    });
  // line-line distance between each interior ring segment and the line
  // if there are no rings this just evaluates to max_float
  let interior_min = other
    .interiors()
    .iter()
    .map(|ring| {
      ring.lines().fold(<T as Bounded>::max_value(), |acc, line| {
        acc.min(self.euclidean_distance(&line))
      })
    })
    .fold(<T as Bounded>::max_value(), |acc, ring_min| {
      acc.min(ring_min)
    });
  // return smaller of the two values
  exterior_min.min(interior_min)
}
```